src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java

```diff
@@ -0,0 +1,51 @@
+package org.commcare.formplayer.services;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.commcare.formplayer.screens.FormplayerQueryScreen;
+import org.commcare.modern.util.Pair;
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cache.CacheManager;
+import org.springframework.cache.annotation.CacheConfig;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.lang.Nullable;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.ByteArrayInputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
+@CacheConfig(cacheNames = {"case_search"})
+public class CaseSearchHelper {
+
+    @Autowired
+    CacheManager cacheManager;
+
+    @Autowired
+    QueryRequester queryRequester;
+
+    @Cacheable(unless="#result == null", key="#username + #uri")
+    public ExternalDataInstance getSearchDataInstance(String username, FormplayerQueryScreen screen, String uri, HttpHeaders headers) {
```

Comment:
Hey @snopoke, I am a bit stuck while implementing caching for Case Search. This method is marked as `@Cacheable` though somehow it's getting executed everytime for the same cache key aka same username and uri passed. Would you be able to take a quick look and see if I am missing something obvious in this implementation.

Reply:
are you testing locally? you need to enable the cache backend: `spring.cache.type=caffeine` (in `application.config`)

Reply:
I think you should also add `case_search` to the list of cache names: https://github.com/dimagi/formplayer/blob/96caf7cb59a09aeb42ba687c36d5ec7364088d11/src/main/resources/application.properties#L33

Reply:
that was it! I was missing both of these params. Thanks a lot.